src/stream/sosemanuk/sosemanuk.c

```diff
@@ -0,0 +1,830 @@
+        x3 ^= ss->kc[(zc) + 3]; \
+    } while (0)
+
+    /*
+     * One Serpent round.
+     *   zc = current subkey counter
+     *   S = S-box macro for this round
+     *   i0 to i4 = input register numbers (the fifth is a scratch register)
+     *   o0 to o3 = output register numbers
+     */
+#define FSS(zc, S, i0, i1, i2, i3, i4, o0, o1, o2, o3)  do { \
+        KA(zc, r ## i0, r ## i1, r ## i2, r ## i3); \
+        S(r ## i0, r ## i1, r ## i2, r ## i3, r ## i4); \
+        SERPENT_LT(r ## o0, r ## o1, r ## o2, r ## o3); \
+    } while (0)
+
+    /*
+     * Last Serpent round. Contrary to the "true" Serpent, we keep
+     * the linear transformation for that last round.
+     */
+#define FSF(zc, S, i0, i1, i2, i3, i4, o0, o1, o2, o3)  do { \
+        KA(zc, r ## i0, r ## i1, r ## i2, r ## i3); \
+        S(r ## i0, r ## i1, r ## i2, r ## i3, r ## i4); \
+        SERPENT_LT(r ## o0, r ## o1, r ## o2, r ## o3); \
+        KA(zc + 4, r ## o0, r ## o1, r ## o2, r ## o3); \
+    } while (0)
+
+    ulong32 r0, r1, r2, r3, r4;
+    unsigned char ivtmp[16];
+
+    if (iv_len >= sizeof ivtmp) {
+        XMEMCPY(ivtmp, iv, sizeof ivtmp);
+    } else {
+        if (iv_len > 0)
+            XMEMCPY(ivtmp, iv, iv_len);
+        XMEMSET(ivtmp + iv_len, 0, (sizeof ivtmp) - iv_len);
+    }
+
+    /*
+     * Decode IV into four 32-bit words (little-endian).
+     */
+    LOAD32L(r0, ivtmp);
+    LOAD32L(r1, ivtmp + 4);
+    LOAD32L(r2, ivtmp + 8);
+    LOAD32L(r3, ivtmp + 12);
+
+    /*
+     * Encrypt IV with Serpent24. Some values are extracted from the
+     * output of the twelfth, eighteenth and twenty-fourth rounds.
+     */
+    FSS(0, S0, 0, 1, 2, 3, 4, 1, 4, 2, 0);
+    FSS(4, S1, 1, 4, 2, 0, 3, 2, 1, 0, 4);
+    FSS(8, S2, 2, 1, 0, 4, 3, 0, 4, 1, 3);
+    FSS(12, S3, 0, 4, 1, 3, 2, 4, 1, 3, 2);
+    FSS(16, S4, 4, 1, 3, 2, 0, 1, 0, 4, 2);
+    FSS(20, S5, 1, 0, 4, 2, 3, 0, 2, 1, 4);
+    FSS(24, S6, 0, 2, 1, 4, 3, 0, 2, 3, 1);
+    FSS(28, S7, 0, 2, 3, 1, 4, 4, 1, 2, 0);
+    FSS(32, S0, 4, 1, 2, 0, 3, 1, 3, 2, 4);
+    FSS(36, S1, 1, 3, 2, 4, 0, 2, 1, 4, 3);
+    FSS(40, S2, 2, 1, 4, 3, 0, 4, 3, 1, 0);
+    FSS(44, S3, 4, 3, 1, 0, 2, 3, 1, 0, 2);
+    ss->s09 = r3;
+    ss->s08 = r1;
+    ss->s07 = r0;
+    ss->s06 = r2;
+
+    FSS(48, S4, 3, 1, 0, 2, 4, 1, 4, 3, 2);
+    FSS(52, S5, 1, 4, 3, 2, 0, 4, 2, 1, 3);
+    FSS(56, S6, 4, 2, 1, 3, 0, 4, 2, 0, 1);
+    FSS(60, S7, 4, 2, 0, 1, 3, 3, 1, 2, 4);
+    FSS(64, S0, 3, 1, 2, 4, 0, 1, 0, 2, 3);
+    FSS(68, S1, 1, 0, 2, 3, 4, 2, 1, 3, 0);
+    ss->r1  = r2;
+    ss->s04 = r1;
+    ss->r2  = r3;
+    ss->s05 = r0;
+
+    FSS(72, S2, 2, 1, 3, 0, 4, 3, 0, 1, 4);
+    FSS(76, S3, 3, 0, 1, 4, 2, 0, 1, 4, 2);
+    FSS(80, S4, 0, 1, 4, 2, 3, 1, 3, 0, 2);
+    FSS(84, S5, 1, 3, 0, 2, 4, 3, 2, 1, 0);
+    FSS(88, S6, 3, 2, 1, 0, 4, 3, 2, 4, 1);
+    FSF(92, S7, 3, 2, 4, 1, 0, 0, 1, 2, 3);
+    ss->s03 = r0;
+    ss->s02 = r1;
+    ss->s01 = r2;
+    ss->s00 = r3;
+
+    ss->ptr = sizeof ss->buf;
```

Comment:
I know there are some reasons why someone would want `sizeof` to not look like a function, but it's all-over like that, so please an you change all `sizeof foo` into `sizeof(foo)`

@karel-m would it probably make sense to add a check of that to helper.pl?

Reply:
detection added to develop as b7e35e9f